pkg/operator/ceph/object/notification/controller.go

```diff
@@ -0,0 +1,227 @@
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+
+	// create the notification if needed
+	if err = r.updateCephBucketNotifications(cephBN); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to create bucket notifications")
+	}
+	// Return and do not requeue
+	logger.Debug("CephBucketNotification done reconciling")
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileNotifications) updateCephBucketNotifications(cephBN *cephv1.CephBucketNotification) error {
+	// step 1: fetch all OBCs that has a label matching this CBN
+	namespace := cephBN.Namespace
+	namespaceListOpt := client.InNamespace(namespace)
+	labelListOpt := client.MatchingLabels{
+		notificationLabelPrefix + cephBN.Name: cephBN.Name,
+	}
+	cephOBCList := &bktv1alpha1.ObjectBucketClaimList{}
+	err := r.client.List(context.TODO(), cephOBCList, namespaceListOpt, labelListOpt)
+	if err != nil {
+		return err
+	}
+	if len(cephOBCList.Items) == 0 {
+		logger.Debugf("no ObjectbucketClaim associated with notification: %q", cephBN.Name)
+		return nil
+	}
+
+	// step 2: get the topic associated with the notification
+	bucketTopic := &cephv1.CephBucketTopic{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephBN.Spec.Topic}, bucketTopic); err != nil {
+		return errors.Wrapf(err, "failed to retrieve CephBucketTopic %q", cephBN.Spec.Topic)
+	}
+	topicARN, err := topic.GetARN(bucketTopic)
+	if err != nil {
+		return errors.Wrap(err, "failed to get topic ARN")
+	}
+
+	// step 3: loop through all OBCs in the list and get their OB and SC
+	for _, cephOBC := range cephOBCList.Items {
+		cephOB := bktv1alpha1.ObjectBucket{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.ObjectBucketName}, &cephOB); err != nil {
+			return errors.Wrapf(err, "failed to retrieve ObjectBucket %q", cephOBC.Spec.ObjectBucketName)
+		}
+		cephSC := storagev1.StorageClass{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.StorageClassName}, &cephSC); err != nil {
+			return errors.Wrapf(err, "failed to retrieve StorageClass %q", cephOBC.Spec.StorageClassName)
+		}
+
+		// step 4: provision the notification
+		p := Provisioner{
+			Client:      r.client,
+			Context:     r.context,
+			ClusterInfo: r.clusterInfo,
+			ClusterSpec: r.clusterSpec,
+		}
+		if err := p.CreateSession(
+			cephOB.Spec.AdditionalState["cephUser"],
+			types.NamespacedName{
+				Name:      cephSC.Parameters["objectStoreName"],
+				Namespace: cephSC.Parameters["objectStoreNamespace"],
+			}); err != nil {
+			return errors.Wrapf(err, "failed to create session for bucket notification. bucket: %q", cephOB.Name)
+		}
+		err = p.Create(&cephOB, topicARN, cephBN)
+		if err != nil {
+			return errors.Wrapf(err, "failed to provision bucket notification: %q", cephBN.Name)
```

Comment:
We have now figured out the name of the bucket as Ceph knows it. Should we add that context here, or is it known from the error given by `p.Create()`?

```suggestion
			return errors.Wrapf(err, "failed to provision notification for bucket %q", cephOB.Name)
```